implementations/rust/ockam/ockam_vault/src/service/vault_service.rs

```diff
@@ -0,0 +1,132 @@
+use crate::service::CreateSoftwareVaultRequest;
+use crate::storage::FileStorage;
+use crate::Vault;
+use minicbor::encode::Write;
+use minicbor::Decoder;
+use ockam_api::{Error, Method, Request, Response, Status};
+use ockam_core::compat::collections::BTreeMap;
+use ockam_core::compat::io;
+use ockam_core::compat::sync::Arc;
+use ockam_core::{Result, Routed, Worker};
+use ockam_node::Context;
+use rand::random;
+use tracing::trace;
+
+/// Vault Service Worker
+#[derive(Default)]
+pub struct VaultService {
+    vaults: BTreeMap<String, Vault>,
+}
+
+impl VaultService {
+    /// Constructor
+    pub fn new() -> Self {
+        Self {
+            vaults: Default::default(),
+        }
+    }
+
+    async fn create_software_vault(
+        &mut self,
+        request: Option<CreateSoftwareVaultRequest<'_>>,
+    ) -> Result<String> {
+        let id: [u8; 16] = random();
+        let id = hex::encode(id.as_slice());
+
+        use std::path::Path;
+
+        let vault = match request {
+            Some(request) => {
+                let path = Path::new(request.path());
+                let temp_path = Path::new(request.temp_path());
+                let storage = FileStorage::create(path, temp_path).await?;
+                Vault::new(Some(Arc::new(storage)))
+            }
+            None => Vault::new(None),
+        };
+
+        self.vaults.insert(id.clone(), vault);
+
+        Ok(id)
+    }
+}
+
+impl VaultService {
+    async fn on_request<W>(&mut self, data: &[u8], buf: W) -> Result<()>
+    where
+        W: Write<Error = io::Error>,
+    {
+        let mut dec = Decoder::new(data);
+        let req: Request = dec.decode()?;
+
+        trace! {
+            target: "ockam_vault::service",
+            id     = %req.id(),
+            method = ?req.method(),
+            path   = %req.path(),
+            "request"
+        }
+
+        let method = match req.method() {
+            Some(m) => m,
+            None => {
+                let error = Error::new(req.path()).with_message("unknown method");
+                Response::not_implemented(req.id())
+                    .body(error)
+                    .encode(buf)?;
+
+                return Ok(());
+            }
+        };
+
+        match method {
+            Method::Get => {}
+            Method::Post => match req.path_segments::<1>().as_slice() {
+                ["software"] => {
+                    let args = if req.has_body() {
+                        let req = dec.decode::<CreateSoftwareVaultRequest>()?;
+                        Some(req)
+                    } else {
+                        None
+                    };
+                    // TODO: Don't ignore the error
+                    let vault_id = self.create_software_vault(args).await?;
+                    Response::ok(req.id()).body(vault_id).encode(buf)?
```

Comment:
Should we wrap simple types (String, Blob, etc.) into tagged structs?

Reply:
For future extensibility it might be a good idea to put those into structs.